tarantool/msgpack_ext/types/datetime.py

```diff
@@ -0,0 +1,237 @@
+        #
+        # https://github.com/python/cpython/blob/1756ffd66a38755cd45de51316d66266ae30e132/Lib/datetime.py#L2323-L2327 
+        initargs = tzinfo.__getinitargs__()
+        if len(initargs) > DATETIME_TIMEZONE_NAME_POS:
+            return initargs[DATETIME_TIMEZONE_NAME_POS]
+        
+        return None
+
+    if isinstance(tzinfo, datetime.tzinfo):
+        # If custom class, tzinfo is expected to have tzname(dt) method
+        # https://github.com/python/cpython/blob/1756ffd66a38755cd45de51316d66266ae30e132/Lib/datetime.py#L1591
+        return tzinfo.tzname(timestamp)
+    
+    raise ValueError(f'Unsupported timezone type {type(tzinfo)}')
+
+
+def is_abbrev_tz(tzname):
+    return tzname in tt_timezones.timezoneAbbrevInfo
+
+def assert_nonambiguous_tz(tzname, tt_tzinfo, error_class):
+    if (tt_tzinfo['category'] & tt_timezones.TZ_AMBIGUOUS) != 0:
+        raise error_class(f'Failed to create datetime with ambiguous timezone "{tzname}"')
+
+def get_python_tzinfo(tzindex):
+    if tzindex not in tt_timezones.indexToTimezone:
+        raise MsgpackError(f'Failed to create datetime with unknown tzindex {tzindex}')
+
+    tzname = tt_timezones.indexToTimezone[tzindex]
+
+    try:
+        tzinfo = pytz.timezone(tzname)
+    except pytz.exceptions.UnknownTimeZoneError:
+        tt_tzinfo = tt_timezones.timezoneAbbrevInfo[tzname]
+        assert_nonambiguous_tz(tzname, tt_tzinfo, MsgpackError)
+
+        tzinfo = datetime.timezone(datetime.timedelta(minutes=tt_tzinfo['offset']),
+                                   name=tzname)
+
+    return tzinfo
+
+def validate_python_timezone(timestamp):
+    tzname = get_python_tzname(timestamp)
+
+    if (tzname is not None):
+        if tzname not in tt_timezones.timezoneToIndex:
+            raise ValueError(f'Failed to create datetime with unknown timezone "{tzname}"')
+
+        if not is_abbrev_tz(tzname):
+            return
+
+        tt_tzinfo = tt_timezones.timezoneAbbrevInfo[tzname]
+        assert_nonambiguous_tz(tzname, tt_tzinfo, ValueError)
+
+def msgpack_decode(data):
+    cursor = 0
+    seconds, cursor = get_bytes_as_int(data, cursor, SECONDS_SIZE_BYTES)
+
+    if len(data) > SECONDS_SIZE_BYTES:
+        nsec, cursor     = get_bytes_as_int(data, cursor, NSEC_SIZE_BYTES)
+        tzoffset, cursor = get_bytes_as_int(data, cursor, TZOFFSET_SIZE_BYTES)
+        tzindex, cursor  = get_bytes_as_int(data, cursor, TZINDEX_SIZE_BYTES)
+    elif len(data) == SECONDS_SIZE_BYTES:
+        nsec = 0
+        tzoffset = 0
+        tzindex = 0
+    else:
+        raise MsgpackError('Unexpected datetime payload length')
+
+    total_nsec = seconds * NSEC_IN_SEC + nsec
+
+    if (tzindex != 0):
+        tzinfo = get_python_tzinfo(tzindex)
+        timestamp = pandas.to_datetime(total_nsec, unit='ns')\
+                          .replace(tzinfo=datetime.timezone.utc)\
+                          .tz_convert(tzinfo)
```

Comment:
tzoffset value from Tarantool may differ from pytz or tt_timezones.timezoneAbbrevInfo[tzname]['offset']. It will be better to use the decoded value.

Reply:
Well, it shouldn't differ. Moreover, there are many complicated timezones which are not simply FixedOffset, so offset is rather useless.

Reply:
Time offset between pytz and Tarantool timezones shouldn't differ since they both are based on Olson database.

In case of tt_timezones.timezoneAbbrevInfo[tzname]['offset'], I think we may add an assert.

Reply:
> Well, it shouldn't differ.

It seems like a user can create Datetime with any offset:
https://www.tarantool.io/en/doc/latest/reference/reference_lua/datetime/new/

Reply:
But it will be `tzoffset` without `tzindex`. It will be okay from our side.

Reply:
Regarding on complicated timezones and why we can't use decoded value.

```
tarantool> dt = datetime.new{year=2008, month=1, day=1, tz='Europe/Moscow'}
---
...

tarantool> dt.tzoffset
---
- 180
...

tarantool> dt2 = dt + datetime.interval.new{month=6}
---
...

tarantool> dt2.tzoffset
---
- 240
...

```

Reply:
It still seems to me that collisions are possible in the future. But let's stop there:

https://github.com/tarantool/tarantool/issues/7680#issuecomment-1246436268